### DISCUSSION TITLE
MAUI Blazor support

### DISCUSSION THREAD

**Discussion by pjmagee** [Megabit/Blazorise#2697] [Ideas]
Ive actively been using Blazorise on the Maui Preview and everything works amazing. Ive come across some minor issues and its to do with Scoped Services. I think for supporting Desktop all that might be needed a slight design change in the configuration when calling `services.Blazorise` and allow uses to toggle `Maui = true|false`. Then based on this option by the user, register the services Scoped or Singleton.

This is my initial thoughts because this is how i got around 1 issue. I am raising it as a discussion here because I dont know the framework too well so maybe there are other things that might not work as expected if using Blazor Control in a Desktop App.

**Comment by stsrki** [comment]
It's not a bad idea. But I'm not sure if it makes sense to add or mention different platforms in a core project, like Maui in this case.

@David-Moreira What do you think? Does it make sense to add this configuration based on a platform?

**Reply by David-Moreira** [reply]
Ah, Well... I don't see why not.
`MAUI` flag would mean we would directly support the best configuration we think it's possible for `MAUI`. That wouldn't be bad as I don't think we'll see `MAUI+Blazor` disappearing anytime soon! haha!

However... there could also be a more generic one that just says `RegisterServicesAsSingleton` which might solve it all the same and not be as bound to a specific platform... 
Or an Enum: `RegisterServicesAs` : Transient / Scoped / Singleton

Other than that, giving the user the flexibility to configure the services `LifeTimeScope` easily is ideal. Yes it might break the app if configured incorrectly... But the user will only ever reach out to this type of configuration to change the blazorise defaults.... if he actually needs it...

**Reply by pjmagee** [reply]
> Ah, Well... I don't see why not.
> `MAUI` flag would mean we would directly support the best configuration we think it's possible for `MAUI`. That wouldn't be bad as I don't think we'll see `MAUI+Blazor` disappearing anytime soon! haha!
> 
> However... there could also be a more generic one that just says `RegisterServicesAsSingleton` which might solve it all the same and not be as bound to a specific platform...
> Or an Enum: `RegisterServicesAs` : Transient / Scoped / Singleton
> 
> Other than that, giving the user the flexibility to configure the services `LifeTimeScope` easily is ideal. Yes it might break the app if configured incorrectly... But the user will only ever reach out to this type of configuration to change the blazorise defaults.... if he actually needs it...

I like the idea of `ServicesAsSingleton` because techincally, its not just MAUI. I believe Blazor WebView2 Components exist for WinForms, WPF, WinUI3 etc. So it could be eventually in the Blasorise documentation you could generalise it as "WebView2 Blazor" usage should use `ServicesAsSingleton`.

**Reply by pjmagee** [reply]
https://github.com/dotnet/aspnetcore/blob/main/src/Components/WebView/WebView/src/Microsoft.AspNetCore.Components.WebView.csproj

![](https://i.imgur.com/CLGoMT9.png)

`.AddBlazoriseWebView(options => ...)` 

or 

`.AddBlazorise(options => { options.WebViewComponent = true } )`

**Reply by stsrki** [reply]
I like the idea of `RegisterServicesAsSingleton`. It is generic and is not bound to any particular platform. And @pjmagee has also a nice idea for an API with `.AddBlazorise(options => { options.WebViewComponent = true } )`.

Considering we currently have mix if singleton and scoped,

```cs
serviceCollection.AddSingleton( configureOptions );
serviceCollection.AddSingleton<BlazoriseOptions>();

serviceCollection.AddSingleton<IIdGenerator, IdGenerator>();
serviceCollection.AddSingleton<IThemeCache, ThemeCache>();
serviceCollection.AddSingleton<IValidationMessageLocalizerAttributeFinder, ValidationMessageLocalizerAttributeFinder>();
serviceCollection.AddScoped<IEditContextValidator, EditContextValidator>();

serviceCollection.AddScoped<ITextLocalizerService, TextLocalizerService>();
serviceCollection.AddScoped( typeof( ITextLocalizer<> ), typeof( TextLocalizer<> ) );

serviceCollection.AddScoped<IValidationHandlerFactory, ValidationHandlerFactory>();
serviceCollection.AddScoped<ValidatorValidationHandler>();
serviceCollection.AddScoped<PatternValidationHandler>();
serviceCollection.AddScoped<DataAnnotationValidationHandler>();
serviceCollection.AddScoped<IMessageService, MessageService>();
serviceCollection.AddScoped<INotificationService, NotificationService>();
serviceCollection.AddScoped<IPageProgressService, PageProgressService>();

serviceCollection.AddSingleton<IDateTimeFormatConverter, DateTimeFormatConverter>();
```

I would name it something along the line, `RegisterScopedAsSingleton`. It is self-descriptive I believe.

```cs
.AddBlazorise(options => { options.RegisterScopedAsSingleton = true } )
```
